### PR TITLE
feat: basic DQN

### LINKED ISSUE #1579: Design the reinforce modal for tradebot
Design the reinforce modal for tradebot to train

### LINKED ISSUE #1580: Train the reinforce modal and tune it
set the trade index for trade bot to train the reinforce modal and tune it

### PR DESCRIPTION
**Pull request by RexBearIU**
### DEVELOP

add env template and modify the year on the footer

### CHECK LIST

- [Naming convention](https://github.com/CAFECA-IO/WorkGuidelines/blob/main/technology/coding-convention/naming-convention.md) verification: checked
- Coding style verification: checked
- new Library: 1 ("@tensorflow/tfjs-node": "^4.17.0")
- new Class / Component: 5
- new loop: 7 (src/strategies/strategy/rl-dqn/train.ts 30-32 and 40 -70 ) (src/price_ticker/price_ticker.service.spec.ts 43-54) (src/strategies/strategies.service.ts 109-112) (src/strategies/strategy/rl-dqn/replay_memory.ts 75-77) (src/strategies/strategy/rl-dqn/tradeAgent.ts 164 and 187)
- new recursive: 0
- high risk: 0
- new sql: 0

### UML

- [Sequence Diagram 001 (interaction between user, frontend and backend)](https://github.com/CAFECA-IO/Documents/blob/main/TBD/TBDSD00001.md)
- [Sequence Diagram 002 (background mechanism)](https://github.com/CAFECA-IO/Documents/blob/main/TBD/TBDSD00002.md)
- [Component Diagram](https://github.com/CAFECA-IO/Documents/blob/main/TBD/TBDCP00001.md)
- [Class Diagram](https://github.com/CAFECA-IO/Documents/blob/main/TBD/TBDCD00001.md)
- [Global context document](https://www.notion.so/d93ac7c941c94fb6b173d34a7bc4e959?pvs=21)
- [Table of interface](https://github.com/CAFECA-IO/EventReport/blob/main/meeting/20230106_tbd_mvp.md)

### CODE REVIEW

File: src/strategies/strategy/rl-dqn/train.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        `cumulativeReward100=${averageReward100.toFixed(1)}; ` +
+          `(epsilon=${agent.epsilon.toFixed(3)}) `,
+      );
+      if (averageReward100 > averageReward100Best) {
+        averageReward100Best = averageReward100;
+        if (savePath != null) {
+          await agent.onlineNetwork.save(
+            'file://src/strategies/strategy/rl-dqn/models/dqn',
+          );
+          console.log(`Saved DQN to ${savePath}`);
```
Comment (Luphia) [inline src/strategies/strategy/rl-dqn/train.ts]: - require deprecate annotation for console.log

File: src/strategies/strategy/rl-dqn/train.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  let averageReward100Best = -Infinity;
+  while (true) {
+    agent.trainOnReplayBatch(batchSize, gamma, optimizer);
+    const { action, cumulativeReward, done } = agent.playStep();
+    // console.log("🚀 ~ action:", action)
+    if (done) {
+      rewardAverager100.append(cumulativeReward);
+      const averageReward100 = rewardAverager100.average();
+
+      console.log(cumulativeReward);
```
Comment (Luphia) [inline src/strategies/strategy/rl-dqn/train.ts]: - require deprecate annotation for console.log

File: src/strategies/strategy/rl-dqn/train.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+      console.log(JSON.stringify(agent.env.profitLog));
```
Comment (Luphia) [inline src/strategies/strategy/rl-dqn/train.ts]: - require deprecate annotation for console.log
- logical operators cannot be used in function parameters

File: src/strategies/strategy/rl-dqn/train.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+      console.log(
```
Comment (Luphia) [inline src/strategies/strategy/rl-dqn/train.ts]: - require deprecate annotation for console.log
- logical operators cannot be used in function parameters

File: src/strategies/strategy/rl-dqn/train.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        JSON.stringify(
+          agent.env.profitLog.reduce((sum, profit) => sum + profit, 0),
+        ),
+      );
+      console.log(
```
Comment (Luphia) [inline src/strategies/strategy/rl-dqn/train.ts]: - require deprecate annotation for console.log
- logical operators cannot be used in function parameters

File: src/strategies/strategy/rl-dqn/train.ts [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+  agent: TradeAgent,
+  batchSize = 32,
+  gamma = 0.99,
+  learningRate = 1e-3,
+  savePath = '/models',
+) {
+  for (let i = 0; i < agent.replayBufferSize; ++i) {
+    agent.playStep();
+  }
+  console.log(agent.cumulativeReward_);
```
Comment (Luphia) [inline src/strategies/strategy/rl-dqn/train.ts]: - require deprecate annotation for console.log

File: src/strategies/strategy/rl-dqn/tradeAgent.ts [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+        const stateTensor = tf.tensor2d(batch.map((example) => example[0]));
+        const actionTensor = tf.tensor1d(
+          batch.map((example) => example[1]),
+          'int32',
+        );
+        let qs = this.onlineNetwork.apply(stateTensor, {
+          training: true,
+        });
+        qs = (qs as tf.Tensor<tf.Rank>)
+          .mul(tf.oneHot(actionTensor, this.NUM_ACTIONS))
```
Comment (Luphia) [inline src/strategies/strategy/rl-dqn/tradeAgent.ts]: - logical operators cannot be used in function parameters

File: src/strategies/strategy/rl-dqn/tradeAgent.ts [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+        action = this.ALL_ACTIONS[actionIndex];
+        // console.log(action);
+      });
+    }
+    const nextState = this.env.step(action);
+    this.replayMemory.append([
+      state,
+      action,
+      nextState.reward,
+      Object.values(nextState).slice(0, 5),
```
Comment (Luphia) [inline src/strategies/strategy/rl-dqn/tradeAgent.ts]: - logical operators cannot be used in function parameters

File: src/strategies/strategy/rl-dqn/tradeAgent.ts [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+      }
+      if (actionNum >= 0.75) {
+        action = 3;
+      }
+    } else {
+      //   Greedily pick an action based on online DQN output.
+      tf.tidy(() => {
+        // const flatState = state.flat(); // Convert state array to a flat array
+        const prediction = this.onlineNetwork.predict(
+          tf.tensor1d(state).reshape([1, -1]),
```
Comment (Luphia) [inline src/strategies/strategy/rl-dqn/tradeAgent.ts]: - logical operators cannot be used in function parameters

File: src/strategies/strategy/rl-dqn/reinforcement-Learning.spec.ts [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+      [0, 1],
+      [1, 0],
+      [1, 1],
+    ]);
+    const ys = tf.tensor2d([[0], [1], [1], [0]]);
+
+    // Train the model
+    model.fit(xs, ys, { epochs: 100 }).then(() => {
+      // Make predictions
+      const prediction = model.predict(tf.tensor2d([[0, 1]]));
```
Comment (Luphia) [inline src/strategies/strategy/rl-dqn/reinforcement-Learning.spec.ts]: - logical operators cannot be used in function parameters

File: src/strategies/strategy/rl-dqn/reinforcement-Learning.spec.ts [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+describe('RL test', () => {
+  it('should be run', () => {
+    // Create a sequential model
+    const model = tf.sequential();
+
+    // Add layers to the model
+    model.add(
+      tf.layers.dense({ units: 2, inputShape: [2], activation: 'relu' }),
+    );
+    model.add(tf.layers.dense({ units: 1, activation: 'sigmoid' }));
```
Comment (Luphia) [inline src/strategies/strategy/rl-dqn/reinforcement-Learning.spec.ts]: - logical operators cannot be used in function parameters

File: src/strategies/strategy/rl-dqn/environment.ts [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+      console.log(
+        'wcnb',
+        this.current_step,
+        profit,
+        this.currentPrice,
+        this.openPrice,
+      );
+    }
+    if (profit < -50) {
+      console.log(
```
Comment (Luphia) [inline src/strategies/strategy/rl-dqn/environment.ts]: - require deprecate annotation for console.log

File: src/strategies/strategy/rl-dqn/environment.ts [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+          reward += 15;
+        } else {
+          reward += -15;
+        }
+      } else {
+        reward = -20;
+      }
+    }
+    if (profit > 50) {
+      console.log(
```
Comment (Luphia) [inline src/strategies/strategy/rl-dqn/environment.ts]: - require deprecate annotation for console.log

File: src/strategies/strategy/rl-dqn/environment.ts [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+        } else {
+          reward += -15;
+        }
+      } else {
+        reward = -20;
+      }
+    }
+    if (action === 2) {
+      if (this.holdingStatus === 0) {
+        this.openPrice = JSON.parse(JSON.stringify(this.currentPrice));
```
Comment (Luphia) [inline src/strategies/strategy/rl-dqn/environment.ts]: - logical operators cannot be used in function parameters

File: src/strategies/strategy/rl-dqn/environment.ts [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+      if (this.holdingStatus === 2) {
+        profit = this.openPrice - (this.currentPrice + this.spreadFee);
+        reward = profit;
+        reward += 10;
+        this.holdingStatus = 0;
+      }
+    }
+    if (action === 1) {
+      if (this.holdingStatus === 0) {
+        this.openPrice = JSON.parse(JSON.stringify(this.currentPrice));
```
Comment (Luphia) [inline src/strategies/strategy/rl-dqn/environment.ts]: - logical operators cannot be used in function parameters

**Review by Luphia — CHANGES_REQUESTED** [review]

File: src/strategies/strategy/rl-dqn/tradeAgent.ts (hunk 7 above)
Comment (RexBearIU) [reply src/strategies/strategy/rl-dqn/tradeAgent.ts]: extract the logical operators

**Review by Luphia — APPROVED** [review]
- no bug found

**Pull request merged**